File: Code/ExecutionEnvironments/Javascript/Application/public/javascripts/models/Executable.js
const { body, oneOf } = require('express-validator')

class Executable {
    #code = "";
    #unitTests = "";
    #executeTests =  false;

    constructor(code, unitTests, executeTests){
        this.#code = code;
        this.#unitTests = unitTests;
        this.#executeTests = executeTests
    }

    get code() {
        return this.#code;
    }

    get unitTests() {
        this.#unitTests;
    }

    get executeTests() {
        this.#executeTests
    }

    static validate = () => {
        return [
            body('code', "value must be provided")
                .exists(),
            body('code', "value can't be empty")
                .not()
                .isEmpty(),
            body('code', 'value must be a string')
                .isString(),
            body("executeTests", "value must be provided")
                .exists(),
            body("executeTests", "value can't be empty")
                .not()
                .isEmpty(),
            body("executeTests", "value must be a boolean")
                .isBoolean(),
            oneOf(
                [
                    body("executeTests").isIn("false"),
                    body("unitTests")
                        .exists()
                        .not()
                        .isEmpty()
                        .isString()
                ],
                "Unit tests must be provided"
                )
        ]
    }
}

module.exports = Executable

File: Code/ExecutionEnvironments/Python/iselearning/models/Executable.py
# Model received by the controller
class Executable:
    def __init__(self, code, unitTests, executeTests):
        self.code = code
        self.unitTests = unitTests
        self.executeTests = executeTests

    @property
    def code(self):
        return self.__code

    @code.setter
    def code(self, code):
        if code is None:
            raise TypeError("Value code must be provided.")
        self.__code = code

    @property
    def unitTests(self):
        return self.__unitTests

    @unitTests.setter
    def unitTests(self, unitTests):
        self.__unitTests = unitTests

    @property
    def executeTests(self):
        return self.__executeTests

    @executeTests.setter
    def executeTests(self, executeTests):
        if not isinstance(executeTests, bool):
            raise TypeError("Value executeTests must be a bool.")
        self.__executeTests = executeTests
